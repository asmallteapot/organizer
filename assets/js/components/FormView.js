import React from 'react'
import axios from 'axios'
import { titles } from '../TitleManager'
import { Form, Text, NestedForm, FormInput } from 'react-form'
import PlacesAutocomplete from 'react-places-autocomplete'
import _ from 'underscore'
import ReactMarkdown from 'react-markdown'
import { csrftoken } from '../Django'
import { Link } from 'react-router-dom'

import FormFieldForm from './FormFieldForm'

export default class FormView extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      form: {
        id: 0,
        title: '',
        description: '',
        fields: []
      },
      submitted: false,
      serverError: undefined
    };
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentDidMount() {
    this.reload();
  }

  reload() {
    this.setState({form: {id:0, title: '', description: '', fields: []}, submitted: false});
    axios.get('/api/forms/'+this.props.match.params.id+'/')
      .then((results) => {
        console.log(results);
        this.setState({form: results.data});
        titles.setTitle(results.data.action.name, results.data.title);
      });
  }

  handleSubmit(values) {
    console.log('submit!', values);
    var data = {
      name: values.name,
      email: values.email,
      address: values.address
    };
    for(var fieldID in values.fields) {
      if (fieldID) {
        data['input_'+fieldID] = values.fields[fieldID];
      }
    }
    axios.post('/api/forms/'+this.props.match.params.id+'/submit_response/',
      data, {headers: {'X-CSRFToken': csrftoken}})
      .then((r) => {
        this.setState({submitted: true});
      })
      .catch((error) => {
        if (error.response.status == 400) {
          var errors = {};
          _.each(
            error.response.data.errors,
            (value, key) => {
               errors[key] = value.join(' ');
            }
          );
          this._form.setAllTouched(true, {errors: errors});
        } else {
          this.setState({serverError: "Recieved "+error.response.status+" from server. Try again."});
        }
      });
  }

  render() {
    if (this.state.submitted) {
      return (
        <div>
          <p>Thanks! Your response has been recorded.</p>
          <p><a onClick={() => {this.reload()}}>Submit another</a></p>
        </div>
      )
    } else {
      return (
        <div>
          <p>{this.state.form.title}</p>
          <ReactMarkdown source={this.state.form.description} />
          <p className="error">{this.state.serverError}</p>
          <Form ref={(r) => {this._form = r}} onSubmit={this.handleSubmit} >
            {({ submitForm, setAllTouched}) => {
              this.setAllTouched = setAllTouched;
              return (
                <form method="post" onSubmit={submitForm}>
                  <label>Email <Text type='text' field='email' /></label>
                  <label>Name <Text type='text' field='name' /></label>
                  <label>
                    Address
                    <FormInput field='address'>
                      {({ setValue, getValue, setTouched }) => {
                        const inputProps = {
                          value: getValue(),
                          onChange: val => setValue(val),
                          onBlur: () => setTouched()
                        }
                        return (
                          <PlacesAutocomplete inputProps={inputProps} />
                        )
                      }}
                    </FormInput>
                  </label>
                  <NestedForm field='fields'>
                    <FormFieldForm fields={this.state.form.fields} />
                  </NestedForm>
                  <input type="submit" className="button" value="Submit" />
                </form>
              )
            }}
          </Form>
        </div>
      )
    }
  }
}
