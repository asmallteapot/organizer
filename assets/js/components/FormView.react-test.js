import React from 'react'
import FormView, { SignupForm, Thanks } from './FormView'
import { Form } from 'react-form'
import { mount } from 'enzyme'
import moment from 'moment'

test('Smoke test', () => {
  const match = {
    params: {
      id: 0
    }
  };
  const component = mount(<FormView match={match} />);

  let tree = component.html();
  expect(tree).toMatchSnapshot();
})

test('Render test', () => {
  const match = {
    params: {
      id: 0
    }
  };
  global.Raven = {
    captureBreadcrumb: jest.fn()
  };
  global.google = {
    maps: {
      places: {
        AutocompleteService: () => ({
          getPlacePredictions: jest.fn(),
        }),
        PlacesServiceStatus: {
          OK: 0
        }
      }
    }
  }
	global.INLINE_FORM_DATA = {
    "fields": [],
    "action": {
      "name": "Berkeley Did Something Stupid, Again",
      "date": moment("20111031", "YYYYMMDD").format()
    },
    "title": "Sign up for an action",
    "description": "Description of an action!",
    "url": "http://organizing.eastbayforward.org/api/forms/5/"
  };

  const component = mount(<FormView match={match} />);

  let tree = component.html();
  expect(tree).toMatchSnapshot();
});

test('Submit values test', () => {
	global.INLINE_FORM_DATA = {
    "fields": [],
    "action": {
      "name": "Berkeley Did Something Stupid, Again",
      "date": moment("20111031", "YYYYMMDD").format()
    },
    "title": "Sign up for an action",
    "description": "Description of an action!",
    "url": "http://organizing.eastbayforward.org/api/forms/5/"
  };
  const fields = [];
  const onSubmit = jest.fn();
  const component = mount(
    <Form onSubmit={onSubmit}>
      {({submitForm}) => (
        <SignupForm fields={fields} onSubmit={submitForm}/>
      )}
    </Form>
  );

  const instance = component.instance();
  const htmlForm = component.find('form').first();
  instance.setValue('email', 'email');
  instance.setValue('name', 'name');
  instance.setValue('address', 'address');
  instance.submitForm();

  const submission = onSubmit.mock.calls[0][0];
  expect(submission).toMatchSnapshot();
})

test('Shows thanks after submit', () => {
  const component = mount(<FormView />);
  expect(component.find(Thanks)).toHaveLength(0);
  component.setState({submitted: true});
  expect(component.find(Thanks)).toHaveLength(1);
  component.find(Thanks).find('a').first().simulate('click')
  expect(component.find(Thanks)).toHaveLength(0);
});
