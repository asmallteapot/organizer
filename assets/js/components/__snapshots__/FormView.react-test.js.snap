// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render test 1`] = `"<div class=\\"expanded row the-form\\"><div class=\\"small-12 columns medium-7 medium-offset-1 the-ask\\"><h1>Sign up for an action</h1><div class=\\"body\\"><div><p><!-- react-text: 13 -->Description of an action!<!-- /react-text --></p></div></div><p class=\\"error\\"></p><div class=\\"ReactForm\\"><form method=\\"post\\"><label><!-- react-text: 18 -->Email <!-- /react-text --><span class=\\"required\\">*</span><div class=\\"FormInput\\"><input type=\\"text\\" value=\\"\\"><div class=\\"FormError\\" style=\\"display: none;\\"></div></div></label><label><!-- react-text: 24 -->Name <!-- /react-text --><span class=\\"required\\">*</span><div class=\\"FormInput\\"><input type=\\"text\\" value=\\"\\"><div class=\\"FormError\\" style=\\"display: none;\\"></div></div></label><label><!-- react-text: 30 -->Address<!-- /react-text --><div class=\\"FormInput\\"><div id=\\"PlacesAutocomplete__root\\" class=\\"\\" style=\\"position: relative; padding-bottom: 0px;\\"><input type=\\"text\\" class=\\"\\" style=\\"display: inline-block; width: 100%; padding: 10px;\\"></div><div class=\\"FormError\\" style=\\"display: none;\\"></div></div></label><div class=\\"FormInput\\"><div class=\\"ReactForm\\"></div></div><input type=\\"submit\\" class=\\"button\\" value=\\"Submit\\"></form></div></div><div class=\\"medium-3 columns meta\\"><div class=\\"name\\">Berkeley Did Something Stupid, Again</div><div class=\\"date\\">October 31st 2011, 12:00:00 am</div><div class=\\"react-add-to-calendar\\"><div class=\\"react-add-to-calendar__wrapper\\"><a class=\\"react-add-to-calendar__button\\"><span><!-- react-text: 45 -->Add to My Calendar <!-- /react-text --><i class=\\"react-add-to-calendar__icon--right fa fa-caret-down\\"></i></span></a></div></div><div class=\\"until\\">6 years ago</div></div></div>"`;

exports[`Smoke test 1`] = `"<div class=\\"spinner\\"><div class=\\"rect1\\"></div><div class=\\"rect2\\"></div><div class=\\"rect3\\"></div><div class=\\"rect4\\"></div><div class=\\"rect5\\"></div></div>"`;

exports[`Submit values test 1`] = `
Object {
  "address": "address",
  "email": "email",
  "name": "name",
}
`;
